### PR TITLE
Added Device Update

### PR DESCRIPTION
**Pull request by beingmattlevy**
Key add was for setDeviceAttributes.
Refactored to keep within the same patterns used.

### CODE REVIEW

**Comment by beingmattlevy** [comment]
We can squash all commits into one clean if desired.

File: lib/leanplum_api/version.rb [hunk 1]
```diff
@@ -1,3 +1,3 @@
 module LeanplumApi
-  VERSION = '3.0.3'
+  VERSION = '3.0.4'
```
Comment (apurvis) [inline lib/leanplum_api/version.rb]: if you change the signature of `track_multi`, you have to do a major version bump, which i'm not sure i endorse, honestly.

File: lib/leanplum_api/api.rb [hunk 2]
```diff
@@ -1,58 +1,53 @@
...
-    # Set the :force_anomalous_override option to catch warnings from leanplum about anomalous events and force them to
-    # not be considered anomalous.
-    def track_multi(events = nil, user_attributes = nil, options = {})
-      request_data = Array.wrap(user_attributes).map { |h| build_user_attributes_hash(h) } +
-                     Array.wrap(events).map { |h| build_event_attributes_hash(h, options) }
+    # This method is for tracking events and/or updating user attributes
+    # at the same time, batched together like leanplum recommends.
+    # Set the :force_anomalous_override option to catch warnings from leanplum
+    # Ôabout anomalous events and force them to not be considered anomalous.
+    def track_multi(events: nil, user_attributes: nil, device_attributes: nil, options: {})
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: though maybe i just don't understand this newfangled syntax?  can i still call 
```
track_multi(events, user_attributes)
```
without the keys?  or does it have to be
```ruby
track_multi(events: events, user_attributes: user_attributes)
```

File: lib/leanplum_api/api.rb [hunk 3]
```diff
@@ -1,58 +1,53 @@
...
+      events = Array.wrap(events)
+
+      request_data = []
+      if user_attributes.present?
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: you don't need this conditional.  a `nil` will just not add anything to the `request_data`
```ruby
irb(main):002:0> Array.wrap(nil)
=> []
```

File: lib/leanplum_api/api.rb [hunk 4]
```diff
@@ -1,58 +1,53 @@
...
+        user_attributes = Array.wrap(user_attributes)
+        user_request_data = user_attributes.map { |h| build_user_attributes_hash(h) }
+        request_data.concat(user_request_data)
+      end
+
+      if device_attributes.present?
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: or this one

File: lib/leanplum_api/api.rb [hunk 5]
```diff
@@ -1,58 +1,53 @@
...
-      request_data = Array.wrap(user_attributes).map { |h| build_user_attributes_hash(h) } +
-                     Array.wrap(events).map { |h| build_event_attributes_hash(h, options) }
+    # This method is for tracking events and/or updating user attributes
+    # at the same time, batched together like leanplum recommends.
+    # Set the :force_anomalous_override option to catch warnings from leanplum
+    # Ôabout anomalous events and force them to not be considered anomalous.
+    def track_multi(events: nil, user_attributes: nil, device_attributes: nil, options: {})
+      events = Array.wrap(events)
+
+      request_data = []
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: why not just initialize 
```ruby
request_data = events.map { |h| build_event_attributes_hash(h, options) }
```

File: lib/leanplum_api/api.rb (hunk 2 above)
Comment (beingmattlevy) [reply lib/leanplum_api/api.rb]: It requires a change to the caller. I'm going to back this change out as it would require more upstream changes. Its a newer way to call methods with variable interface calls. But I'll keep to the existing pattern for this change.

File: lib/leanplum_api/version.rb (hunk 1 above)
Comment (apurvis) [reply lib/leanplum_api/version.rb]: i endorse the major version bump to get the `device_attributes` in the arguments, actually.  change this to `4.0.0`

also need information in `CHANGELOG.md`

File: README.md [hunk 6]
```diff
@@ -82,7 +82,7 @@ api.track_events(event)
 api.track_events(event, allow_offline: true)
 
 # You can also track events and user attributes at the same time
-api.track_multi(event, attribute_hash)
+api.track_multi(events: event, user_attributes: attribute_hash)
```
Comment (apurvis) [inline README.md]: you need to update the documentation for the new `device_attributes` key, and you also need to update the line below about how to use the `force_anomalous_override` option.

File: lib/leanplum_api/api.rb (hunk 2 above)
Comment (beingmattlevy) [reply lib/leanplum_api/api.rb]: Hmmm... I do need to add device_attributes to the interface. I did not like the idea of forcing the use of options or type check to distinguish between user and device attributes.

File: README.md (hunk 6 above)
Comment (beingmattlevy) [reply README.md]: will do

File: leanplum_api.gemspec [hunk 7]
```diff
@@ -5,24 +5,26 @@ require 'leanplum_api/version'
...
   gem.test_files  = Dir["spec/**/*"]
 
-  gem.required_ruby_version = '>= 1.9.3'
+  gem.required_ruby_version = '>= 2.3'
 
   gem.add_dependency 'activesupport', '> 3.0', '< 5'
   gem.add_dependency 'awesome_print', '~> 1'
-  gem.add_dependency 'faraday', '~> 0.9', '>= 0.9.1'
-  gem.add_dependency 'faraday_middleware', '~> 0.10'
+  gem.add_dependency 'faraday', '~> 0.12', '>= 0.12.2'
```
Comment (apurvis) [inline leanplum_api.gemspec]: i can't accept this change; our app currently has faraday 0.9.2 and so this will be uninstallable

File: leanplum_api.gemspec [hunk 8]
```diff
@@ -5,24 +5,26 @@ require 'leanplum_api/version'
...
+  gem.email       = ['analytics-dev@lumoslabs.com', 'eng@countable.us']
   gem.homepage    = 'http://www.github.com/lumoslabs/leanplum_api'
   gem.summary     = 'Gem for the Leanplum API'
   gem.description = 'Ruby-esque access to Leanplum API'
   gem.licenses    = ['MIT']
   gem.files       = Dir["lib/**/*"] + ['Gemfile', 'LICENSE.txt', 'README.md']
   gem.test_files  = Dir["spec/**/*"]
 
-  gem.required_ruby_version = '>= 1.9.3'
+  gem.required_ruby_version = '>= 2.3'
```
Comment (apurvis) [inline leanplum_api.gemspec]: agree we should bump this but why does it need to be 2.3?  what feature does 2.3 have that makes it required?

File: lib/leanplum_api/api.rb [hunk 9]
```diff
@@ -1,58 +1,53 @@
 module LeanplumApi
   class API
+    EXPORT_PENDING = 'PENDING'.freeze
+    EXPORT_RUNNING = 'RUNNING'.freeze
+    EXPORT_FINISHED = 'FINISHED'.freeze
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: let's not move this up here.  leave it after `extend` and the `class` declaration.

File: lib/leanplum_api/api.rb (hunk 2 above)
Comment (apurvis) [reply lib/leanplum_api/api.rb]: yeah i'm cool with this change, actually... having 3 hash arguments + an options hash is kind of gross.  needs to go to 4.0 anyways so have at it.

File: lib/leanplum_api/api.rb [hunk 10]
```diff
@@ -1,58 +1,53 @@
...
     end
 
-    # This method is for tracking events and/or updating user attributes at the same time, batched together like
-    # leanplum recommends.
-    # Set the :force_anomalous_override option to catch warnings from leanplum about anomalous events and force them to
-    # not be considered anomalous.
-    def track_multi(events = nil, user_attributes = nil, options = {})
-      request_data = Array.wrap(user_attributes).map { |h| build_user_attributes_hash(h) } +
-                     Array.wrap(events).map { |h| build_event_attributes_hash(h, options) }
+    # This method is for tracking events and/or updating user attributes
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: "updating user or device attributes"

File: spec/fixtures/vcr/track_events.yml [hunk 11]
```diff
@@ -2,11 +2,11 @@
 http_interactions:
 - request:
     method: post
-    uri: https://www.leanplum.com/api?action=multi&apiVersion=1.0.6&appId=<LEANPLUM_APP_ID>&clientKey=<LEANPLUM_PRODUCTION_KEY>&devMode=false&time=1502510400
+    uri: https://www.leanplum.com/api?action=multi&apiVersion=1.0.6&appId=<LEANPLUM_APP_ID>&clientKey=<LEANPLUM_PRODUCTION_KEY>&devMode=false&time=1502521200
     body:
       encoding: UTF-8
-      string: '{"data":[{"action":"track","event":"purchase","userId":123456,"time":"1502510400","params":{"some_timestamp":"2015-05-01
-        01:02:03"}},{"action":"track","event":"purchase_page_view","userId":54321,"time":"1502509800"}]}'
+      string: '{"data":[{"action":"track","event":"purchase","userId":123456,"time":"1502521200","params":{"some_timestamp":"2015-05-01
```
Comment (apurvis) [inline spec/fixtures/vcr/track_events.yml]: this should not change - right now if i run these specs it will fail, because of my time zone, which sucks.  i used `Timecop` to freeze the spec run time, but maybe it needs an option about which time zone to use or something... either way this change should not be necessary.

File: lib/leanplum_api/api.rb (hunk 5 above)
Comment (apurvis) [reply lib/leanplum_api/api.rb]: you can just do all of this in a few lines
```ruby
request_data = Array.wrap(events).map { |h| build_event_attributes_hash(h, options) }
             + Array.wrap(user_attributes).map { |h| build_user_attributes_hash(h) }
             + Array.wrap(device_attributes).map { |h| build_device_attributes_hash(h) }
```
might also save some resources depending on how `Array#concat` handles memory allocation

File: spec/api_spec.rb [hunk 12]
```diff
@@ -13,6 +13,46 @@
...
     }]
+   end
+
+  let(:devices) do
+    [{
+       device_id: 'fu123',
+       appVersion: 'x42x',
+       deviceModel: 'p0d',
+       create_date: '2018-01-01'.to_date
+     }]
```
Comment (apurvis) [inline spec/api_spec.rb]: line up the whitespace please

File: spec/fixtures/vcr/set_device_attributes.yml [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+      - Google Frontend
+      Content-Length:
+      - '121'
+      Expires:
+      - Thu, 31 Aug 2017 17:33:45 GMT
+      Cache-Control:
+      - private
+    body:
+      encoding: UTF-8
+      string: '{"response":[{"success":true,"warning":{"message":"Device clock skew
```
Comment (apurvis) [inline spec/fixtures/vcr/set_device_attributes.yml]: are we sure "Device clock skew" is a reasonable warning?

File: spec/spec_helper.rb [hunk 14]
```diff
@@ -3,6 +3,13 @@
 require 'timecop'
 require 'webmock'
 require 'vcr'
+require 'dotenv/load'
+
+# ENV['LEANPLUM_PRODUCTION_KEY']='prod_I02AwsAsRGi7g0NNc2YA2lyUceOROrkDLDHcheWBNk4'
+# ENV['LEANPLUM_APP_ID']='app_bQZEF70IGUuPPYSLcZEON5nlihNa9N9GukO93mSpT48'
```
Comment (apurvis) [inline spec/spec_helper.rb]: you probably shouldn't commit valid credentials to a public repo....

**Comment by apurvis** [comment]
Can you please also delete this method when you are bumping the major version number?

https://github.com/lumoslabs/leanplum_api/blob/1f523e2626f25494b064d5980e5043f1c7b6cb5b/lib/leanplum_api/api.rb#L124-L128

just remove all that.

File: spec/fixtures/vcr/track_events.yml (hunk 11 above)
Comment (apurvis) [reply spec/fixtures/vcr/track_events.yml]: https://github.com/travisjeffery/timecop/issues/41
https://github.com/travisjeffery/timecop/issues/182

File: lib/leanplum_api/api.rb (hunk 5 above)
Comment (beingmattlevy) [reply lib/leanplum_api/api.rb]: 👍

File: lib/leanplum_api/api.rb (hunk 2 above)
Comment (beingmattlevy) [reply lib/leanplum_api/api.rb]: Its actually 2 hashes. or specifically a hash with a hash as value for options key. this is a goods addition to ruby for reducing the need for positional arguments and using keys.

File: leanplum_api.gemspec (hunk 7 above)
Comment (beingmattlevy) [reply leanplum_api.gemspec]: ok. I'll bump it back down.

File: leanplum_api.gemspec (hunk 8 above)
Comment (beingmattlevy) [reply leanplum_api.gemspec]: speed.

File: spec/spec_helper.rb (hunk 14 above)
Comment (beingmattlevy) [reply spec/spec_helper.rb]: oops. I hate it when that happens. I added dotenv gem to repo to handle this.

Comment (apurvis) [reply spec/spec_helper.rb]: feel free to add `dotenv` and tell people how to run the specs with an env file - that's basically what i have done while testing as well.  but you should at this point delete this app from leanplum, because this will never come out of the commit history.

File: leanplum_api.gemspec (hunk 8 above)
Comment (apurvis) [reply leanplum_api.gemspec]: that's not a good answer for people still running apps on 2.x - this is overly restrictive for them.

fine to move it to `>= 2.0` but there's no point in restricting the ruby version arbitrarily beyond that unless you _require_ a given feature introduced in 2.x

File: lib/leanplum_api/api.rb [hunk 15]
```diff
@@ -4,55 +4,42 @@ class API
...
-                     Array.wrap(events).map { |h| build_event_attributes_hash(h, options) }
+    # This method is for tracking events and/or updating user and/or device attributes
+    # at the same time, batched together like leanplum recommends.
+    # Set the :force_anomalous_override option to catch warnings from leanplum
+    # Ôabout anomalous events and force them to not be considered anomalous.
+    def track_multi(events: nil, user_attributes: nil, device_attributes: nil, options: {})
+      events = Array.wrap(events)
+
+      request_data = Array.wrap(events).map { |h| build_event_attributes_hash(h, options) }
+      request_data += Array.wrap(user_attributes).map { |h| build_user_attributes_hash(h) }
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: it's more efficient from a memory allocation standpoint to not use `+=`.  just use `+` as in my original example.

File: lib/leanplum_api/api.rb [hunk 16]
```diff
@@ -210,21 +222,30 @@ def development_connection
 
     # Deletes the user_id and device_id key/value pairs from the hash parameter.
     def extract_user_id_or_device_id_hash!(hash)
-      user_id = hash.delete(:user_id)
-      device_id = hash.delete(:device_id)
+      user_id = hash.delete(:user_id) || hash.delete(:userIid)
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: `userIid`?

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,3 +1,6 @@
+## 4.0.0
+* Added support for setDeviceAttributes
```
Comment (apurvis) [inline CHANGELOG.md]: can you add a note like 
**BREAKING CHANGE** - multi method signature has changed.

File: README.md [hunk 18]
```diff
@@ -81,8 +81,8 @@ api.track_events(event)
 # Events tracked like this will be made part of a session; for independent events use :allow_offline
 api.track_events(event, allow_offline: true)
 
-# You can also track events and user attributes at the same time
-api.track_multi(event, attribute_hash)
+# You can also track events, user attributes, and device attributes at the same time. magic!
+api.track_multi(events: event, user_attributes: user_attributes, device_attributes: device_attributes)
```
Comment (apurvis) [inline README.md]: you also need to change this part of the README
https://github.com/lumoslabs/leanplum_api/blame/master/README.md#L91
to reflect the new method signature

File: lib/leanplum_api/api.rb [hunk 19]
```diff
@@ -210,21 +222,30 @@ def development_connection
...
+      device_id = hash.delete(:device_id) || hash.delete(:deviceId)
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: if you're going to accept `deviceId`, that should be documented in the readme...  not sure why we should accept it though?

File: lib/leanplum_api/api.rb [hunk 20]
```diff
@@ -210,21 +222,32 @@ def development_connection
...
     # Action can be any command that takes a userAttributes param.  "start" (a session) is the other command that most
     # obviously takes userAttributes.
     # As of 2015-10 Leanplum supports ISO8601 date & time strings as user attributes.
-    def build_user_attributes_hash(user_hash, action = 'setUserAttributes')
-      user_hash = HashWithIndifferentAccess.new(user_hash)
-      user_hash.each { |k, v| user_hash[k] = v.iso8601 if v.is_a?(Date) || v.is_a?(Time) || v.is_a?(DateTime) }
+    def build_user_attributes_hash(user_hash)
+      user_hash = fix_iso8601(user_hash)
+      user_attributes_hash = extract_user_id_or_device_id_hash!(user_hash).merge(action: 'setUserAttributes', userAttributes: user_hash)
+      user_attributes_hash[:devices] = user_hash[:devices] if user_hash[:devices].present?
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: why is this necessary?  `user_attributes_hash` is already `user_hash`, just with ISO adjustments, and `user_id` removed, right?

Comment (apurvis) [reply lib/leanplum_api/api.rb]: oh wait, i see.  shouldn't it be.  i think this isn't quite right though, because you need to remove the `:devices` key from the hash _before_ you create the `user_attributes_hash`.   otherwise you've got the `devices:` in 2 places
```
{
  userAttributes: {
    attribute: 'val',
    devices: devices_hash
  },
  devices: devices_hash
}
```

Comment (beingmattlevy) [reply lib/leanplum_api/api.rb]: its about the devices hash.
https://www.leanplum.com/docs/api/production#setuserattributes
devices has to be at the top level, not embedded in userAttributes

File: lib/leanplum_api/api.rb (hunk 19 above)
Comment (beingmattlevy) [reply lib/leanplum_api/api.rb]: leanplum api has deviceId, userId. the gem converts device_id for you, but its confusing to have device_id, deviceModel in the same hash.

File: spec/api_spec.rb [hunk 21]
```diff
@@ -31,6 +71,16 @@
       })
     end
 
+    it 'builds user_attributes_hash with device' do
+      user = users.first.merge(devices: devices)
+      user_attributes_hash = api.send(:build_user_attributes_hash, user)
+      expect(user_attributes_hash[:devices].first).to eq({ device_id: 'fu123',
```
Comment (apurvis) [inline spec/api_spec.rb]: sorry there's no hound on this repo, but this formatting is not ok.
1. shouldn't this just be `devices.first.with_indifferent_access`?
2. if not, format like this
```
expect(user_attributes_hash[:devices].first).to eq({
  device_id: 'fu123',
  appVersion: 'x42x',
  deviceModel: 'p0d',
  create_date: '2018-01-01'.to_date
}.with_indifferent_access)
```

File: lib/leanplum_api/api.rb [hunk 22]
```diff
@@ -210,21 +222,33 @@ def development_connection
...
+    
     # Action can be any command that takes a userAttributes param.  "start" (a session) is the other command that most
     # obviously takes userAttributes.
     # As of 2015-10 Leanplum supports ISO8601 date & time strings as user attributes.
-    def build_user_attributes_hash(user_hash, action = 'setUserAttributes')
-      user_hash = HashWithIndifferentAccess.new(user_hash)
-      user_hash.each { |k, v| user_hash[k] = v.iso8601 if v.is_a?(Date) || v.is_a?(Time) || v.is_a?(DateTime) }
+    def build_user_attributes_hash(user_hash)
+      user_hash = fix_iso8601(user_hash)
+      devices = user_hash.delete(:devices)
```
Comment (apurvis) [inline lib/leanplum_api/api.rb]: please document this feature

**Comment by apurvis** [comment]
getting pretty close, though there's still a few comments that need addressing.  you're still working on this PR, right?

**Comment by beingmattlevy** [comment]
I was on vacation last week. :) Will be making some changes this week. If you have specific formatting changes needed for your organization, please feel free to re-format. I'll try to address the open concerns this week.

**Comment by beingmattlevy** [comment]
Updated with changes. Please feel free to align spacing or re-build vcr cassettes files if/as needed. I'll keep an eye out for changes on your master and provide any changes I have back.

File: README.md (hunk 18 above)
Comment (apurvis) [reply README.md]: this still hasn't been addressed (otherwise looks great)

File: lib/leanplum_api/configuration.rb [hunk 23]
```diff
@@ -29,6 +29,9 @@ class Configuration
     attr_accessor :logger
     attr_accessor :timeout_seconds
 
+    # Override validations for leanplum response. On by default.
+    attr_accessor :validate_response
```
Comment (apurvis) [inline lib/leanplum_api/configuration.rb]: curious - why do we want this option?

File: lib/leanplum_api/version.rb [hunk 24]
```diff
@@ -1,3 +1,3 @@
 module LeanplumApi
-  VERSION = '3.0.3'
+  VERSION = '4.0.3'
```
Comment (apurvis) [inline lib/leanplum_api/version.rb]: 4.0 please

**Comment by apurvis** [comment]
this looks good... unfortunately though, i've got a PR coming up to add a separate feature and there might be some merge damage.  let me open that PR so you can see what's going on.

File: lib/leanplum_api/configuration.rb (hunk 23 above)
Comment (beingmattlevy) [reply lib/leanplum_api/configuration.rb]: I stub LP responses in our Rails app. VCR won't integrate for 'reasons'. Hence not checking counts of return success vs send items. Only a testing feature.

**Comment by apurvis** [comment]
https://github.com/lumoslabs/leanplum_api/pull/18

File: lib/leanplum_api/configuration.rb (hunk 23 above)
Comment (apurvis) [reply lib/leanplum_api/configuration.rb]: :ok:

**Comment by beingmattlevy** [comment]
I don't have time to integrate #18 into my fork at the moment. Please feel free to accept or not these changes. Pull in and overwrite the version as you need. I have already launched with 4.0.2, and 4.0.3.

File: .env.example [hunk 25]
```diff
@@ -0,0 +1,6 @@
+# Example env file. Copy this file to .env and fill with your corresponding LeanPlum key values.
```
Comment (apurvis) [inline .env.example]: does `bundle exec rspec` automatically load this file? that would be dope

File: spec/api_spec.rb [hunk 26]
```diff
@@ -31,6 +71,27 @@
       })
     end
 
+    it 'builds user_attributes_hash with devices' do
+      user = users.first
+      user[:devices] = devices
+      expect(api.send(:build_user_attributes_hash, users.first)).to eq({
+                                                                         userId: first_user_id,
```
Comment (apurvis) [inline spec/api_spec.rb]: re: style - this is _way_ out of style guide compliance.  max line length is 120.

basically you just need to left indent this stuff, or assign it to a variable beforehand.  sorry i can overlook a little whitespace but i can't overlook this.
https://github.com/bbatsov/ruby-style-guide

File: .env.example (hunk 25 above)
Comment (beingmattlevy) [reply .env.example]: it does. 
You can also have .env.development, .env.test, etc.... with env specific overrides from the base .env file.

File: spec/api_spec.rb (hunk 26 above)
Comment (beingmattlevy) [reply spec/api_spec.rb]: Please feel free to make the style work for you. :)

Comment (apurvis) [reply spec/api_spec.rb]: the reason the style guide exists is so that questions like this don't come up.  it is the arbiter, and this is just not how you style ruby.  there's an [entire company/application](https://houndci.com/) devoted to rejecting this type of formatting.

are you willing to let your otherwise well composed PR die on this molehill?

Comment (apurvis) [reply spec/api_spec.rb]: i'll make you a deal - you make this block sane and i'll handle the merge of `build_user_attributes`

Comment (beingmattlevy) [reply spec/api_spec.rb]: done.

File: README.md (hunk 18 above)
Comment (beingmattlevy) [reply README.md]: done

**Comment by beingmattlevy** [comment]
OK. I think I sorted the style changes needed. Let me know if you need anything else.

**Comment by apurvis** [comment]
replaced by https://github.com/lumoslabs/leanplum_api/pull/19

**Comment by apurvis** [comment]
released 4.0
